ehrql/query_engines/base_sql.py

```diff
@@ -77,19 +78,39 @@ def get_query(self, variable_definitions):
         `get_setup_and_cleanup_queries` on the query object.
         """
         variable_definitions = apply_transforms(variable_definitions)
+
+        # Generate a table containing the IDs all of patients matching the population
+        # definition
+        population_expression = self.get_predicate(variable_definitions["population"])
+        select_patient_id = self.select_patient_id_for_population(population_expression)
+        population_query = select_patient_id.where(population_expression)
+        population_query = apply_patient_joins(population_query)
+        population_table = self.reify_query(population_query)
+        # Store a reference to the population table so that we can use it while
+        # generating the variable expressions below
+        self.population_table = population_table
```

Comment:
I understand why you've done this, but making this class non-trivially stateful makes me a little bit uneasy.  This code's already quite demanding to grok, but at least it's more-or-less functional.  That said, I can't think of an improvement on your approach that doesn't involve lots of tedious plumbing, which would likely make the code harder to read.